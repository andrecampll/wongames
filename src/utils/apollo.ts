import { useMemo } from 'react';
import { setContext } from '@apollo/client/link/context';
import {
  ApolloClient,
  HttpLink,
  NormalizedCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { Session } from 'next-auth';
import apolloCache from './apolloCache';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

function createApolloClient(session?: Session | null) {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
  });

  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || '';
    const authorization = jwt ? `Bearer ${jwt}` : '';

    return { headers: { ...headers, authorization } };
  });

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: authLink.concat(httpLink),
    cache: apolloCache,
  });
}

export function initializeApollo(
  initialState = null,
  session?: Session | null,
) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session);

  if (initialState) {
    apolloClientGlobal.cache.restore(
      initialState as NormalizedCacheObject & NormalizedCache,
    );
  }

  if (typeof window === 'undefined') return apolloClientGlobal;

  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = null, session?: Session) {
  const store = useMemo(() => initializeApollo(initialState, session), [
    initialState,
    session,
  ]);

  return store;
}
