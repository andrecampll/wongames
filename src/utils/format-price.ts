export function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('en', {
    style: 'currency',
    currency: 'USD',
  }).format(price);
}
