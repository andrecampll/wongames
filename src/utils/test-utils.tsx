import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';

import {
  WishlistContext,
  WishlistContextDefaultValues,
  WishlistContextDTO,
} from '../hooks/wishlist/useWishlist';

import theme from '../styles/theme';
import {
  CartContext,
  CartContextDTO,
  CartContextDefaultValues,
} from '../hooks/cart/useCart';

type CustomRenderProps = {
  cartProviderProps?: CartContextDTO;
  wishlistProviderProps?: WishlistContextDTO;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        <WishlistContext.Provider value={wishlistProviderProps}>
          {ui}
        </WishlistContext.Provider>
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
