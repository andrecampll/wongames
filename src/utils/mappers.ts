import { QueryGames_games } from '../graphql/generated/QueryGames';
import {
  QueryHome_banners,
  QueryHome_sections_freeGames_highlight,
} from '../graphql/generated/QueryHome';
import { QueryWishlist_wishlists_games } from '../graphql/generated/QueryWishlist';

import { formatPrice } from './format-price';

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map(({ image, title, subtitle, button, ribbon }) => ({
    image: `http://localhost:1337${image?.url}`,
    title,
    subtitle,
    buttonLabel: button?.label,
    buttonLink: button?.link,
    ...(ribbon && {
      ribbon: ribbon.text,
      ribbonColor: ribbon.color,
      ribbonSize: ribbon.size,
    }),
  }));
};

export const gamesMapper = (games: QueryGames_games[] | null | undefined) => {
  return games
    ? games.map(({ id, name, slug, developers, cover, price }) => ({
        id,
        title: name,
        slug,
        developer: developers[0].name,
        image: cover?.url
          ? `http://localhost:1337${cover?.url}`
          : 'https://i.stack.imgur.com/y9DpT.jpg',
        price,
      }))
    : [];
};

export const highlightMapper = (
  highlight: QueryHome_sections_freeGames_highlight | null | undefined,
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment,
      }
    : {};
};

export const cartMapper = (
  games:
    | QueryGames_games[]
    | QueryWishlist_wishlists_games[]
    | null
    | undefined,
) => {
  return games
    ? games.map(game => ({
        id: game.id,
        img: `http://localhost:1337${game.cover?.url}`,
        title: game.name,
        price: formatPrice(game.price),
      }))
    : [];
};
