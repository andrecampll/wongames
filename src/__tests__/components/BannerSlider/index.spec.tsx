/* eslint-disable import/no-unresolved */
import 'match-media-mock';
import { screen } from '@testing-library/react';
import BannerSlider from '../../../components/BannerSlider';
import { render } from '../../../utils/test-utils';

const items = [
  {
    image: 'https://source.unsplash.com/user/willianjusten/1042x580',
    title: 'Defy death 1',
    subtitle: '<p>Play the new <strong>CrashLands</strong> season',
    buttonLabel: 'Buy now',
    buttonLink: '/games/defy-death',
    ribbon: 'Bestselling',
  },
  {
    image: 'https://source.unsplash.com/user/willianjusten/1042x582',
    title: 'Defy death 2',
    subtitle: '<p>Play the new <strong>CrashLands</strong> season',
    buttonLabel: 'Buy now',
    buttonLink: '/games/defy-death',
  },
];

describe('BannerSlider', () => {
  it('should be able to render vertical slider', () => {
    const { container } = render(<BannerSlider items={items} />);

    expect(container.querySelector('.slick-vertical')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able to render with 1 active item', () => {
    const { container } = render(<BannerSlider items={items} />);

    expect(container.querySelectorAll('.slick-slide')).toHaveLength(2);
    expect(container.querySelectorAll('li.slick-active')).toHaveLength(1);

    expect(
      screen.getByRole('heading', { name: /defy death 1/i, hidden: false }),
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: /defy death 2/i, hidden: true }),
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able to render vertical slider', () => {
    const { container } = render(<BannerSlider items={items} />);

    expect(container.querySelector('.slick-dots')).toBeInTheDocument();
  });
});
