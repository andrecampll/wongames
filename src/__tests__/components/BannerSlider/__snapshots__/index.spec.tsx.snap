// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BannerSlider should be able to render vertical slider 1`] = `
<section
  class="sc-fubCfw gURrpr"
>
  <section
    class="sc-bdfBwQ iqTTGR"
  >
    <div
      class="slick-slider slick-vertical slick-initialized"
      dir="ltr"
    >
      <div
        class="slick-list"
        style="height: 0px;"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-hKgILt hNAidq"
              >
                <div
                  class="sc-dlfnbm cnqZPj"
                  color="primary"
                >
                  Bestselling
                </div>
                <div
                  aria-label="Defy death 1"
                  class="sc-eCssSg idfjjV"
                  role="img"
                  src="https://source.unsplash.com/user/willianjusten/1042x580"
                />
                <div
                  class="sc-jSgupP bdRVCZ"
                >
                  <h2
                    class="sc-gKsewC fZqpCl"
                  >
                    Defy death 1
                  </h2>
                  <h3
                    class="sc-iBPRYJ ermzdX"
                  >
                    <p>
                      Play the new 
                      <strong>
                        CrashLands
                      </strong>
                       season
                    </p>
                  </h3>
                  <a
                    class="sc-gsTCUz cohTjt"
                    href="/games/defy-death"
                  >
                    <span>
                      Buy now
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-hKgILt hNAidq"
              >
                <div
                  aria-label="Defy death 2"
                  class="sc-eCssSg hfjqkj"
                  role="img"
                  src="https://source.unsplash.com/user/willianjusten/1042x582"
                />
                <div
                  class="sc-jSgupP bdRVCZ"
                >
                  <h2
                    class="sc-gKsewC fZqpCl"
                  >
                    Defy death 2
                  </h2>
                  <h3
                    class="sc-iBPRYJ ermzdX"
                  >
                    <p>
                      Play the new 
                      <strong>
                        CrashLands
                      </strong>
                       season
                    </p>
                  </h3>
                  <a
                    class="sc-gsTCUz cohTjt"
                    href="/games/defy-death"
                  >
                    <span>
                      Buy now
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="slick-dots"
        style="display: block;"
      >
        <li
          class="slick-active"
        >
          <button>
            1
          </button>
        </li>
        <li
          class=""
        >
          <button>
            2
          </button>
        </li>
      </ul>
    </div>
  </section>
</section>
`;

exports[`BannerSlider should be able to render with 1 active item 1`] = `
<section
  class="sc-fubCfw gURrpr"
>
  <section
    class="sc-bdfBwQ iqTTGR"
  >
    <div
      class="slick-slider slick-vertical slick-initialized"
      dir="ltr"
    >
      <div
        class="slick-list"
        style="height: 0px;"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-hKgILt hNAidq"
              >
                <div
                  class="sc-dlfnbm cnqZPj"
                  color="primary"
                >
                  Bestselling
                </div>
                <div
                  aria-label="Defy death 1"
                  class="sc-eCssSg idfjjV"
                  role="img"
                  src="https://source.unsplash.com/user/willianjusten/1042x580"
                />
                <div
                  class="sc-jSgupP bdRVCZ"
                >
                  <h2
                    class="sc-gKsewC fZqpCl"
                  >
                    Defy death 1
                  </h2>
                  <h3
                    class="sc-iBPRYJ ermzdX"
                  >
                    <p>
                      Play the new 
                      <strong>
                        CrashLands
                      </strong>
                       season
                    </p>
                  </h3>
                  <a
                    class="sc-gsTCUz cohTjt"
                    href="/games/defy-death"
                  >
                    <span>
                      Buy now
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                class="sc-hKgILt hNAidq"
              >
                <div
                  aria-label="Defy death 2"
                  class="sc-eCssSg hfjqkj"
                  role="img"
                  src="https://source.unsplash.com/user/willianjusten/1042x582"
                />
                <div
                  class="sc-jSgupP bdRVCZ"
                >
                  <h2
                    class="sc-gKsewC fZqpCl"
                  >
                    Defy death 2
                  </h2>
                  <h3
                    class="sc-iBPRYJ ermzdX"
                  >
                    <p>
                      Play the new 
                      <strong>
                        CrashLands
                      </strong>
                       season
                    </p>
                  </h3>
                  <a
                    class="sc-gsTCUz cohTjt"
                    href="/games/defy-death"
                  >
                    <span>
                      Buy now
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <ul
        class="slick-dots"
        style="display: block;"
      >
        <li
          class="slick-active"
        >
          <button>
            1
          </button>
        </li>
        <li
          class=""
        >
          <button>
            2
          </button>
        </li>
      </ul>
    </div>
  </section>
</section>
`;
